code/game/machinery/vendors/custom_vendors.dm

```diff
@@ -42,33 +65,82 @@
+					bag.remove_from_storage(used)
 				used.moveToNullspace()
-				user.visible_message("<span class='notice'>[user] puts [used] into [src].</span>", "<span class='notice>'You put [used] into [src].</span>")
-				return ITEM_INTERACT_COMPLETE
+				if(!quiet)
+					user.visible_message("<span class='notice'>[user] puts [used] into [src].</span>", "<span class='notice>'You put [used] into [src].</span>")
+				return INSERT_DONE
+
+	if(!can_wait)
+		return INSERT_NEEDS_INPUT
 
 	var/price = tgui_input_number(user, "How much do you want to sell [used] for?")
 	if(!isnum(price))
-		return ITEM_INTERACT_COMPLETE
+		return INSERT_FAIL
 	if(!Adjacent(user))
 		to_chat(user, "<span class='warning'>You can't reach [src] from here!</span>")
-		return ITEM_INTERACT_COMPLETE
-	if(!user.is_holding(used))
-		to_chat(user, "<span class='warning'>\The [used] isn't in your hand anymore!</span>")
-		return ITEM_INTERACT_COMPLETE
-	if(!user.canUnEquip(used, FALSE))
-		to_chat(user, "<span class='warning'>\The [used] is stuck to your hand!</span>")
-		return ITEM_INTERACT_COMPLETE
+		return INSERT_FAIL
+	if(isnull(bag))
+		if(!user.is_holding(used))
+			to_chat(user, "<span class='warning'>\The [used] isn't in your hand anymore!</span>")
+			return INSERT_FAIL
+		if(!user.canUnEquip(used, FALSE))
+			to_chat(user, "<span class='warning'>\The [used] is stuck to your hand!</span>")
+			return INSERT_FAIL
+	else
+		if(!user.is_holding(bag))
+			to_chat(user, "<span class='warning'>\The [bag] isn't in your hand anymore!</span>")
+			return INSERT_FAIL
+		if(used.loc != bag)
+			to_chat(user, "<span class='warning'>\The [used] isn't in [bag] anymore!</span>")
+			return INSERT_FAIL
 
 	var/datum/data/vending_product/physical/record = new(used.name, used.icon, used.icon_state)
 	record.items += used
 	record.price = price
 	physical_product_records += record
 	SStgui.update_uis(src, TRUE)
-	user.unequip(used)
+	if(isnull(bag))
+		user.unequip(used)
+	else
+		bag.remove_from_storage(used)
 	used.moveToNullspace()
-	user.visible_message("[user] puts [used] into [src].", "You put [used] into [src].")
-	return ITEM_INTERACT_COMPLETE
+	if(!quiet)
+		user.visible_message("<span class='notice'>[user] puts [used] into [src].</span>", "<span class='notice'>You put [used] into [src].</span>")
+	return INSERT_DONE
+
+/obj/machinery/economy/vending/custom/MouseDrop_T(atom/dragged, mob/user, params)
+	if(!istype(dragged, /obj/item/storage))
+		return ..()
+
+	var/obj/item/storage/bag = dragged
+	var/inserted = FALSE
+	for(var/obj/item/thing in bag.contents.Copy())
+		var/result = try_add_stock(user, thing, can_wait = FALSE, quiet = TRUE, bag = bag)
+		if(result == INSERT_FAIL)
+			break
+		if(result == INSERT_DONE)
+			inserted = TRUE
+			continue
+
+		// result == INSERT_NEEDS_INPUT
+		if(inserted)
+			user.visible_message("<span class='notice'>[user] transfers some things from [bag] into [src].</span>", "<span class='notice'>You transfer some things from [bag] into [src].</span>")
```

Comment:
Maybe add a blind message to these, too?

Reply:
I thought about it, but I don't think it'd actually make enough noise to be recognizable as anything. With all the other stuff that's surely making noise on the station, I doubt you'd notice it at all.